plugins/incremental-ingestion-backend/src/engine/IncrementalIngestionEngine.ts

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { DeferredEntity } from '@backstage/plugin-catalog-backend';
+import {
+  INCREMENTAL_ENTITY_PROVIDER_ANNOTATION,
+  IterationEngine,
+  IterationEngineOptions,
+} from '../types';
+import { IncrementalIngestionDatabaseManager } from '../database/IncrementalIngestionDatabaseManager';
+import type { AbortSignal } from 'node-abort-controller';
+
+import { performance } from 'perf_hooks';
+import { Duration, DurationObjectUnits } from 'luxon';
+import { v4 } from 'uuid';
+
+export class IncrementalIngestionEngine implements IterationEngine {
+  restLength: Duration;
+  backoff: DurationObjectUnits[];
+
+  private manager: IncrementalIngestionDatabaseManager;
+
+  constructor(private options: IterationEngineOptions) {
+    this.manager = options.manager;
+    this.restLength = Duration.fromObject(options.restLength);
+    this.backoff = options.backoff ?? [
+      { minutes: 1 },
+      { minutes: 5 },
+      { minutes: 30 },
+      { hours: 3 },
+    ];
+  }
+
+  async taskFn(signal: AbortSignal) {
+    try {
+      this.options.logger.debug('Begin tick');
+      await this.handleNextAction(signal);
+    } catch (error) {
+      this.options.logger.error(`${error}`);
+      throw error;
+    } finally {
+      this.options.logger.debug('End tick');
+    }
+  }
+
+  async handleNextAction(signal: AbortSignal) {
+    await this.options.ready;
+
+    const result = await this.getCurrentAction();
+    if (result) {
```

Comment:
matter of taste, so optional: could consider inverting this check and logging and returning, so that we don't get so much extra indentation of the "meat" of the method

Reply:
I think we're going to leave this as-is for the moment.